### ISSUE TITLE
The producer/consumer howto could really do with some examples

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3835]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#3835|
|Type|enhancement|
|Created|2009-05-13 11:42:42Z|
|Branch|https://github.com/twisted/twisted/tree/streaming-example-3835|

The current version of the producer/consumer howto is more of a reference. Some examples involving transports would make it easier to assimilate. It's a bit problematic though since transports are both most common consumers and most common source of data for producers, which might be confusing. But the examples should at least show some transport integration.

<details><summary>Searchable metadata</summary>

```
trac-id__3835 3835
type__enhancement enhancement
reporter__itamar itamar
priority__normal normal
milestone__ 
branch__branches_streaming_example_3835 branches/streaming-example-3835
branch_author__ruttbe__thijs ruttbe, thijs
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation documentation
time__1242214962000000 1242214962000000
changetime__1302542958000000 1302542958000000
version__None None
owner__thijs thijs
cc__thijs cc__ruttbe
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @thijstriemstra|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_3"></a>|@thijstriemstra commented|
|-|-|

[#4171](https://github.com/twisted/twisted/issues/4171) was a duplicate of this and contains an example.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_4"></a>|@thijstriemstra commented|
|-|-|

(In [28001]) Branching to 'streaming-example-3835'

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_6"></a>|@thijstriemstra commented|
|-|-|

(In [28002]) Add streaming.py example, refs [#3835](https://github.com/twisted/twisted/issues/3835), [#4171](https://github.com/twisted/twisted/issues/4171)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_7"></a>|@thijstriemstra commented|
|-|-|

(In [30518]) apply coding standard, add to examples listing and news file. refs [#3835](https://github.com/twisted/twisted/issues/3835)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra removed owner<br>@thijstriemstra set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @thijstriemstra|
|-|-|

  1. `cnt` is a silly variable name.  Please switch to `count`.
  1. There's lots of extra parenthesis where `print` is used.
  1. The example doesn't actually require Python 2.6, it works fine on 2.4.  The last sentence of the docstring can probably just go.
  1. `random.randrange` is preferred over `random.randint`
  1. `ServeRandom` should use `sendLine` instead of `transport.write` in `connectionMade`
  1. The `pauseProducing` docstring is nice.  How about some for the other methods?
  1. `stopProducing` should cause the `resumeProducing` loop to stop, too.
  1. The vertical whitespace needs some cleanup.
  1. The producer/consumer howto should link to this example.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_11"></a>|@thijstriemstra commented|
|-|-|

(In [30760]) address review comments, refs [#3835](https://github.com/twisted/twisted/issues/3835)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_12"></a>|@thijstriemstra removed owner<br>@thijstriemstra set status to `new`|
|-|-|

Up for 500 points.. :)

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_13"></a>|@mithrandi set owner to @thijstriemstra|
|-|-|

This example only implements a push producer, which is somewhat different from a pull producer. I think the various references to "example of a producer" (or similar wording) should explicitly mention that this is a push producer, to avoid confusion.

Aside from that, everything else looks good; please go ahead and merge once that issue is dealt with.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_15"></a>|@thijstriemstra commented|
|-|-|

(In [31631]) add references to push producer, refs [#3835](https://github.com/twisted/twisted/issues/3835)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_16"></a>|@thijstriemstra commented|
|-|-|

(In [31632]) reference push producer, refs [#3835](https://github.com/twisted/twisted/issues/3835)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_17"></a>|@thijstriemstra set status to `closed`|
|-|-|

(In [31633]) Added sample implementation of a Twisted push producer/consumer system.

Author: ruttbe, thijs
Reviewer: exarkun, mithrandi
Fixes: [#3835](https://github.com/twisted/twisted/issues/3835)
